source/extensions/filters/network/zookeeper_proxy/decoder.cc

```diff
@@ -476,41 +476,101 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
 }
 
 Network::FilterStatus DecoderImpl::onData(Buffer::Instance& data) {
-  return decodeAfterBuffer(data, DecodeType::READ);
+  return decodeAndBuffer(data, zk_filter_read_buffer_, DecodeType::READ);
 }
 
 Network::FilterStatus DecoderImpl::onWrite(Buffer::Instance& data) {
-  return decodeAfterBuffer(data, DecodeType::WRITE);
+  return decodeAndBuffer(data, zk_filter_write_buffer_, DecodeType::WRITE);
 }
 
-Network::FilterStatus DecoderImpl::decodeAfterBuffer(Buffer::Instance& data, DecodeType dtype) {
+Network::FilterStatus DecoderImpl::decodeAndBuffer(Buffer::Instance& data,
+                                                   Buffer::OwnedImpl& zk_filter_buffer,
+                                                   DecodeType dtype) {
+  uint32_t data_len = data.length();
+  uint32_t zk_filter_buffer_len = zk_filter_buffer.length();
+
```

Comment:
add comments to state the reason we prepend and drop

Reply:
Added comments in L500-502